app/com/arpnetworking/rollups/ConsistencyChecker.java

```diff
@@ -0,0 +1,579 @@
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.CompletionException;
+import java.util.function.Consumer;
+import javax.inject.Inject;
+import javax.inject.Named;
+
+/**
+ * Actor that compares rollup datapoints to their source material, and logs any discrepancies.
+ *
+ * @author Spencer Pearson (spencerpearson at dropbox dot com)
+ */
+public class ConsistencyChecker extends AbstractActorWithTimers {
+
+    private final KairosDbClient _kairosDbClient;
+    private final MetricsFactory _metricsFactory;
+    private final ActorRef _queue;
+
+    @Override
+    public Receive createReceive() {
+        return new ReceiveBuilder()
+                .match(Task.class, this::startTask)
+                .match(SampleCounts.class, this::sampleCountsReceived)
+                .match(CollectionActor.QueueEmpty.class, msg ->
+                    getTimers().startSingleTimer("WORK_REQUEST_BACKOFF", WORK_REQUEST_BACKOFF_EXPIRED_MSG, REQUEST_WORK_BACKOFF)
+                )
+                .matchEquals(WORK_REQUEST_BACKOFF_EXPIRED_MSG, msg -> this.requestWork())
+                .build();
+    }
+
+    /**
+     * {@link ConsistencyChecker} actor constructor.
+     *
+     * @param kairosDbClient kairosdb client
+     * @param metricsFactory metrics factory
+     * @param queue queue to request {@link Task}s from
+     */
+    @Inject
+    public ConsistencyChecker(
+            final KairosDbClient kairosDbClient,
+            final MetricsFactory metricsFactory,
+            @Named("RollupConsistencyCheckerQueue") final ActorRef queue
+    ) {
+        _kairosDbClient = kairosDbClient;
+        _metricsFactory = metricsFactory;
+        _queue = queue;
+    }
+
+    @Override
+    public void preStart() throws Exception {
+        super.preStart();
+        requestWork();
+    }
+
+    private void requestWork() {
+        _queue.tell(CollectionActor.Poll.getInstance(), getSelf());
+    }
+
+    private void startTask(final Task task) {
+        Patterns.pipe(
+                _kairosDbClient.queryMetrics(buildCountComparisonQuery(task))
+                        .thenApply(response -> {
+                            try {
+                                return ConsistencyChecker.parseSampleCounts(task, response);
+                            } catch (final MalformedSampleCountResponse err) {
+                                throw new CompletionException(err);
+                            }
+                        })
+                        .whenComplete((response, failure) -> requestWork()),
+                getContext().dispatcher()
+        ).to(getSelf());
+    }
+
+    private void sampleCountsReceived(final SampleCounts sampleCounts) {
+        final Task task = sampleCounts.getTask();
+        final Optional<Throwable> failure = sampleCounts.getFailure();
+
+        final Metrics metrics = _metricsFactory.create();
```

Comment:
This should probably be in a try with resources to ensure it's closed even on exception.